From 7feff74c3b24757674776ccdf6e66068ef60a09c Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Mon, 22 Oct 2018 15:19:27 +0800
Subject: [PATCH 5001/5242] MMFMWK-8297: ISI: return error when ISI chan0 is
 busy

commit  e85213feec92aef9a636a980d58af020bd8fdb81 from
https://source.codeaurora.org/external/imx/linux-imx.git

When ISI channel0 is used to recevice camera data, it can
not be used to mem2mem. So return -EBUSY error code when
user try to use them at the same time.

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
(cherry picked from commit c6d5bf866819deaaf4900ae47d575c5a36d30a66)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-isi-cap.c |    5 +++++
 drivers/media/platform/imx8/mxc-isi-m2m.c |    5 +++++
 2 files changed, 10 insertions(+)

diff --git a/drivers/media/platform/imx8/mxc-isi-cap.c b/drivers/media/platform/imx8/mxc-isi-cap.c
index 4498793..7d9ae38 100644
--- a/drivers/media/platform/imx8/mxc-isi-cap.c
+++ b/drivers/media/platform/imx8/mxc-isi-cap.c
@@ -633,6 +633,11 @@ static int mxc_isi_capture_open(struct file *file)
 
 	dev_dbg(&mxc_isi->pdev->dev, "%s, ISI%d\n", __func__, mxc_isi->id);
 
+	if (mxc_isi->is_m2m) {
+		v4l2_err(mxc_isi->v4l2_dev, "%s: ISI channel[%d] is busy\n",
+					__func__, mxc_isi->id);
+		return -EBUSY;
+	}
 	atomic_inc(&mxc_isi->open_count);
 	mxc_isi->is_m2m = 0;
 
diff --git a/drivers/media/platform/imx8/mxc-isi-m2m.c b/drivers/media/platform/imx8/mxc-isi-m2m.c
index 900a323..47f1def 100644
--- a/drivers/media/platform/imx8/mxc-isi-m2m.c
+++ b/drivers/media/platform/imx8/mxc-isi-m2m.c
@@ -396,6 +396,11 @@ static int mxc_isi_m2m_open(struct file *file)
 
 	dev_dbg(dev, "%s, ISI%d\n", __func__, mxc_isi->id);
 
+	if (atomic_read(&mxc_isi->open_count) > 0) {
+		dev_err(dev, "%s: ISI channel[%d] is busy\n", __func__, mxc_isi->id);
+		return -EBUSY;
+	}
+
 	if (mutex_lock_interruptible(&mxc_isi->lock))
 		return -ERESTARTSYS;
 	mxc_ctx = kzalloc(sizeof(*mxc_ctx), GFP_KERNEL);
-- 
1.7.9.5

